gpuav: Update debug label regions management

arno-lunarg opened the pull request:
Took some of what Artem did for syncval
Add debug label regions for debug printf
Still need to add support for secondary command buffers, not so easy

EDIT: Updated PR to add support for debug label regions starting in one (primary) command buffer and ending in another following one. To do that I drew inspiration from sync val, but had to come up with my own (little) system, for GPU-AV has different constraints. I described the problem in a comment above this new `vvl::CommandBuffer` method: https://github.com/KhronosGroup/Vulkan-ValidationLayers/blob/19390298820aceb3881e9865f05eabdd27cecf41/layers/state_tracker/cmd_buffer_state.h#L710

EDIT 2: Reworked again to account for Artem's comments. Removed debug label region tracking from the `CommandBuffer` class, now it is tracked on a "per command buffer submission" basis, in the `CommandBufferSubmission` struct. Added new test `StageInfoWithDebugLabel7` that was not working without this enhancement. 
Realized that support for debug printf was broken (https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/8924) so removed it

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 305756.

arno-lunarg commented on layers/gpu/instrumentation/gpuav_instrumentation.cpp:
```diff
@@ -395,42 +395,23 @@ void PreCallSetupShaderInstrumentationResources(Validator &gpuav, CommandBuffer
+    std::string debug_region_name;
+    if (!cb_state.GetLabelCommands().empty()) {
+        const uint32_t last_label_command_i = static_cast<uint32_t>(cb_state.GetLabelCommands().size() - 1);
+        debug_region_name = cb_state.GetDebugRegionName(cb_state.GetLabelCommands(), last_label_command_i);
```
This is how I use `GetDebugRegionName`, anything looking wrong @artem-lunarg ?

artem-lunarg replied:
Looks good. Not sure what is the scope in gpu-av. If gpu-av needs also to consider global label state then `GetDebugRegionName()` provides  `initial_label_stack` parameter and it can be used to add that support to.

Labels command influence is not limited to command buffer in general. EndLabel can be in different command buffer. So after specific queue submission the queue can have current non-empty label name (per-queue state). It can be used as `initial_label_stack` to prepend when replaying command from current command buffer.

arno-lunarg replied:
Ultimately we want the same level of support as sync val, so this is valuable info! I need to look at how you do those things and will probably ping you again on this topic

artem-lunarg replied:
Just refreshed memory how queue label state is tracked. It's vvl::Queue and `cmdbuf_label_stack` field.
Example how it is used in syncval: https://github.com/KhronosGroup/Vulkan-ValidationLayers/blob/fdbdd359fb2bbb50352e84b259171813957d1ee3/layers/sync/sync_validation.cpp#L3202

One difference of this queue state form command buffer label tracking that on queue level it's enough to keep std::vector<string> of nested labels, in contrast to LabelCommand in the command buffer. The reason is because after command buffer is submitted we know exact label hierarchy (just replay label commands from entire submitted command buffer) so we just save result of this replay.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 18062 running.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 18062 failed.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 306110.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 18069 failed.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 306142.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 18070 running.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 18070 failed.

artem-lunarg reviewed (approved):
LGTM if CI is happy

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 306976.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 18088 running.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 18088 passed.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 308710.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 18108 running.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 18108 passed.

artem-lunarg commented on layers/state_tracker/cmd_buffer_state.h:
```diff
@@ -698,6 +699,15 @@ class CommandBuffer : public RefcountedStateObject {
+    // to construct an initial label stack. sync-val can do that because validation and error reporting is done
+    // *before* a command buffer list is submitted: validation is performed one command buffer at a time,
+    // and `GetQueueState()->cmdbuf_label_stack` is updated between those validations.
+    // When GPU-AV starts doing error reporting, the label stack info stored in Queue state is lost.
```
@arno-lunarg could you provide more details on the following parts, to understand this better

> GPU-AV cannot just look at `GetQueueState()->cmdbuf_label_stack`

> // When GPU-AV starts doing error reporting, the label stack info stored in Queue state is lost.

From what I remember, Queue tracks label stack globally, it reflects state after the labels from the previously submitted command buffers are applied. And then when we start processing new command buffer then we use global queue state to and call it the initial label state (at the beginning of current command buffer). It's not clear how the point (Pre/Post) that was used to update Queue can affect processing of subsequent command buffer.

artem-lunarg replied:
> the label stack info stored in Queue state is lost.

This also not clear. Label stack in Queue is never lost. It reflects current queue state after all past submissions.

artem-lunarg replied:
Had an internal discussion. Arno explained how GPU-AV reads results, and it's different timeline then where we track queue progress.

artem-lunarg commented on layers/state_tracker/state_tracker.cpp:
```diff
@@ -1264,7 +1264,8 @@ void ValidationStateTracker::PostCallRecordQueueSubmit(VkQueue queue, uint32_t s
         for (uint32_t i = 0; i < submit->commandBufferCount; i++) {
-            if (auto cb_state = GetRead<vvl::CommandBuffer>(submit->pCommandBuffers[i])) {
+            if (auto cb_state = GetWrite<vvl::CommandBuffer>(submit->pCommandBuffers[i])) {
+                cb_state->SetPostProcessInitialLabelStack(queue_state->cmdbuf_label_stack);
```
Related to the previous comment, this place is not obvious. I guess in general we don't update command buffers during QueueSubmit since they can be shared.

arno-lunarg commented:
Had a call with Artem, maintaining some state per command buffer looks wrong. And state needs to be maintained per command buffer _submission_, not per command buffer: the same command buffer could be submitted multiple times, so maintaining some state inside a command buffer would have it overwritten when looping through the submitted command buffers

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 309825.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 309842.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 18126 running.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 18126 passed.

arno-lunarg commented:
EDIT 2: Reworked again to account for Artem's comments. Removed debug label region tracking from the `CommandBuffer` class, now it is tracked on a "per command buffer submission" basis, in the `CommandBufferSubmission` struct. Added new test `StageInfoWithDebugLabel7` that was not working without this enhancement.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 309999.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 18127 running.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 18127 failed.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 310032.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 18129 passed.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 310096.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 18130 running.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 18130 passed.

artem-lunarg commented on layers/state_tracker/queue_state.h:
```diff
@@ -34,11 +35,33 @@ namespace vvl {
+    // When GPU-AV starts doing error reporting, when command buffers have completed,
+    // the label stack info stored in Queue state is lost.
+    // => GPU-AV needs to track this initial label stack per command buffer submission.
+    std::vector<std::string> initial_label_stack;
```
As a future improvement I wonder if this can be put into GPU-AV subclasses, to minimize changes to core types.

artem-lunarg commented on layers/state_tracker/state_tracker.cpp:
```diff
@@ -1198,38 +1214,39 @@ void ValidationStateTracker::PreCallRecordQueueSubmit(VkQueue queue, uint32_t su
         }
 
-        for (uint32_t i = 0; i < submit->signalSemaphoreCount; ++i) {
+        for (uint32_t sig_sem_i = 0; sig_sem_i < submit->signalSemaphoreCount; ++sig_sem_i) {
```
I can't say it's easier to read this rename, I actually starting to spend time to decode sig_sem_i sequence. IMO it's valuable property or very nested scopes is that they allow to use shorter names.

arno-lunarg replied:
Yes true, as long as it is not `index` or `idx` I am happy!

artem-lunarg reviewed (approved):
LGTM. One future suggestion to consider if it's possible to store initial label stack in parallel GPU-AV specific structures.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 311063.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 18145 running.

arno-lunarg commented:
Update: Rename `wait_sem_i` and `sig_sem_i` to `i`, replaced some `auto` usage in `queue_state.cpp`

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 311095.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 18146 failed.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 312397.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build queued with queue ID 312588.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 18153 running.

ci-tester-lunarg commented:
CI Vulkan-ValidationLayers build # 18153 passed.
